File: packages/dotnet/test/Codec.Net.Tests/ToolWatcherTests.cs
// SPDX-License-Identifier: MIT
//
// ToolWatcher tests: mirror packages/web/test/tool-watcher.test.ts and
// packages/python/tests/test_tool_watcher.py.
using System;
using System.Collections.Generic;
using Xunit;

namespace Codec.Tests;

public class ToolWatcherTests
{
    private const uint START = 90;
    private const uint END   = 91;

    private static TokenizerMap SynthMap() => new()
    {
        Id = "test/synth",
        Version = "2",
        VocabSize = 100,
        Encoder = "byte_level",
        Vocab = new() { ["hello"] = 0, ["world"] = 1, ["!"] = 2, ["foo"] = 3, ["bar"] = 4 },
        SpecialTokens = new() { ["<tool_call>"] = 90, ["</tool_call>"] = 91 },
    };

    [Fact]
    public void PassthroughThenRegionThenPassthrough()
    {
        var w = new ToolWatcher(SynthMap(), "<tool_call>", "</tool_call>");
        // "hello world <tool_call> foo bar </tool_call> hello !"
        var evs = w.Feed(new uint[] { 0, 1, START, 3, 4, END, 0, 2 });
        Assert.Equal(3, evs.Count);

        Assert.Equal(WatcherEventKind.Passthrough, evs[0].Kind);
        Assert.Equal(new uint[] { 0, 1 }, evs[0].Ids);

        Assert.Equal(WatcherEventKind.Region, evs[1].Kind);
        Assert.Equal(new uint[] { 3, 4 }, evs[1].Ids);  // markers excluded

        Assert.Equal(WatcherEventKind.Passthrough, evs[2].Kind);
        Assert.Equal(new uint[] { 0, 2 }, evs[2].Ids);

        Assert.False(w.Inside);
    }

    [Fact]
    public void RegionSplitAcrossFeeds()
    {
        var w = new ToolWatcher(SynthMap(), "<tool_call>", "</tool_call>");

        // Feed 1: region opens but doesn't close.
        var evs = w.Feed(new uint[] { 0, START, 3 });
        Assert.Single(evs);
        Assert.Equal(WatcherEventKind.Passthrough, evs[0].Kind);
        Assert.Equal(new uint[] { 0 }, evs[0].Ids);
        Assert.True(w.Inside);

        // Feed 2: closes the region with body accumulated across both feeds.
        evs = w.Feed(new uint[] { 4, END, 1 });
        Assert.Equal(2, evs.Count);
        Assert.Equal(WatcherEventKind.Region, evs[0].Kind);
        Assert.Equal(new uint[] { 3, 4 }, evs[0].Ids);
        Assert.Equal(WatcherEventKind.Passthrough, evs[1].Kind);
        Assert.Equal(new uint[] { 1 }, evs[1].Ids);
        Assert.False(w.Inside);
    }

    [Fact]
    public void MultipleRegionsInOneFeed()
    {
        var w = new ToolWatcher(SynthMap(), "<tool_call>", "</tool_call>");
        var evs = w.Feed(new uint[] { 0, START, 3, END, 1, START, 4, END, 2 });
        Assert.Equal(5, evs.Count);
        Assert.Equal(WatcherEventKind.Passthrough, evs[0].Kind);
        Assert.Equal(WatcherEventKind.Region,      evs[1].Kind);
        Assert.Equal(new uint[] { 3 }, evs[1].Ids);
        Assert.Equal(WatcherEventKind.Passthrough, evs[2].Kind);
        Assert.Equal(WatcherEventKind.Region,      evs[3].Kind);
        Assert.Equal(new uint[] { 4 }, evs[3].Ids);
        Assert.Equal(WatcherEventKind.Passthrough, evs[4].Kind);
    }

    [Fact]
    public void StrayEndPassesThrough()
    {
        var w = new ToolWatcher(SynthMap(), "<tool_call>", "</tool_call>");
        var evs = w.Feed(new uint[] { 0, END, 1 });
        // End with no preceding start: ordinary token.
        Assert.Single(evs);
        Assert.Equal(WatcherEventKind.Passthrough, evs[0].Kind);
        Assert.Equal(new uint[] { 0, END, 1 }, evs[0].Ids);
    }

    [Fact]
    public void MissingSpecialNameThrows()
    {
        Assert.Throws<ToolWatcherException>(
            () => new ToolWatcher(SynthMap(), "<not_real>", "</tool_call>"));
    }

    [Fact]
    public void ResetDropsInFlightRegion()
    {
        var w = new ToolWatcher(SynthMap(), "<tool_call>", "</tool_call>");
        w.Feed(new uint[] { START, 3, 4 });
        Assert.True(w.Inside);
        w.Reset();
        Assert.False(w.Inside);
        // End marker now becomes a stray (no buffered body).
        var evs = w.Feed(new uint[] { END, 1 });
        Assert.Single(evs);
        Assert.Equal(WatcherEventKind.Passthrough, evs[0].Kind);
        Assert.Equal(new uint[] { END, 1 }, evs[0].Ids);
    }

    /// <summary>
    /// No-decode contract: mirror of test_watcher_does_not_decode_tokens
    /// in libcodec / @codecai/web / codecai. Use a map with empty vocab
    /// and feed IDs outside any plausible vocab range. The watcher must
    /// emit them verbatim.
    /// </summary>
    [Fact]
    public void NeverDecodesOperatesOnRawIds()
    {
        var noVocab = new TokenizerMap
        {
            Id = "test/no-vocab",
            Version = "2",
            VocabSize = 4,
            Encoder = "byte_level",
            Vocab = new(),
            SpecialTokens = new() { ["<tool_call>"] = 90, ["</tool_call>"] = 91 },
        };
        var w = new ToolWatcher(noVocab, "<tool_call>", "</tool_call>");
        const uint BIG_A = 0xFFFFFF00, BIG_B = 0xDEADBEEF, BIG_C = 0xCAFEBABE;
        var evs = w.Feed(new uint[] { 12345, BIG_A, START, BIG_B, BIG_C, END, 99999 });
        Assert.Equal(3, evs.Count);
        Assert.Equal(new uint[] { 12345, BIG_A }, evs[0].Ids);
        Assert.Equal(new uint[] { BIG_B, BIG_C }, evs[1].Ids);  // body verbatim
        Assert.Equal(new uint[] { 99999 }, evs[2].Ids);
    }

    [Fact]
    public void IntOverloadAcceptsWireFrameType()
    {
        // CodecFrame.Ids is int[]; verify the int overload works.
        var w = new ToolWatcher(SynthMap(), "<tool_call>", "</tool_call>");
        var evs = w.Feed(new int[] { 0, 1, (int)START, 3, (int)END, 2 });
        Assert.Equal(3, evs.Count);
        Assert.Equal(WatcherEventKind.Region, evs[1].Kind);
        Assert.Equal(new uint[] { 3 }, evs[1].Ids);
    }

    // ── Ordering: interleaved events in stream order (defect 3) ──────────────
    //
    // [a, S, X, E, b, S, Y, E, c] must produce five ORDERED events:
    // passthrough(a) / region(X) / passthrough(b) / region(Y) / passthrough(c).
    // This is the exact shape every language's watcher must agree on.

    [Fact]
    public void OrderingMatchesDefect3Example()
    {
        var w = new ToolWatcher(SynthMap(), "<tool_call>", "</tool_call>");
        const uint a = 0, b = 1, c = 2, x = 3, y = 4; // hello, world, !, foo, bar
        var evs = w.Feed(new uint[] { a, START, x, END, b, START, y, END, c });
        Assert.Equal(5, evs.Count);
        Assert.Equal(WatcherEventKind.Passthrough, evs[0].Kind);
        Assert.Equal(new uint[] { a }, evs[0].Ids);
        Assert.Equal(WatcherEventKind.Region, evs[1].Kind);
        Assert.Equal(new uint[] { x }, evs[1].Ids);
        Assert.Equal(WatcherEventKind.Passthrough, evs[2].Kind);
        Assert.Equal(new uint[] { b }, evs[2].Ids);
        Assert.Equal(WatcherEventKind.Region, evs[3].Kind);
        Assert.Equal(new uint[] { y }, evs[3].Ids);
        Assert.Equal(WatcherEventKind.Passthrough, evs[4].Kind);
        Assert.Equal(new uint[] { c }, evs[4].Ids);
    }

    // ── Nested start markers (defect 5) ───────────────────────────────────────

    [Fact]
    public void NestedStartIsDroppedFromBodyButObservable()
    {
        var w = new ToolWatcher(SynthMap(), "<tool_call>", "</tool_call>");
        // S 0 S 1 E 2 -> NestedStart / Region([0,1]) / Passthrough([2])
        var evs = w.Feed(new uint[] { START, 0, START, 1, END, 2 });
        Assert.Equal(3, evs.Count);
        Assert.Equal(WatcherEventKind.NestedStart, evs[0].Kind);
        Assert.Equal(new uint[] { START }, evs[0].Ids);
        Assert.Equal(WatcherEventKind.Region, evs[1].Kind);
        Assert.Equal(new uint[] { 0, 1 }, evs[1].Ids);
        Assert.Equal(WatcherEventKind.Passthrough, evs[2].Kind);
        Assert.Equal(new uint[] { 2 }, evs[2].Ids);
    }

    // ── Truncation: End() while inside a region (defect 1) ────────────────────
    //
    // An unterminated region (stream ends mid tool-call, e.g. the model hit
    // its length limit) used to be silently dropped: no event, no signal,
    // indistinguishable from a model that never called a tool. End() must
    // report it, carrying the finish reason so a length stop is
    // distinguishable from a malformed emission.

    [Fact]
    public void EndEmitsTruncatedWithFinishReason()
    {
        var w = new ToolWatcher(SynthMap(), "<tool_call>", "</tool_call>");
        var evs = w.Feed(new uint[] { 0, START, 3, 4 });
        Assert.Single(evs);
        Assert.Equal(WatcherEventKind.Passthrough, evs[0].Kind);
        Assert.True(w.Inside);

        evs = w.End("length");
        Assert.Single(evs);
        Assert.Equal(WatcherEventKind.Truncated, evs[0].Kind);
        Assert.Equal(new uint[] { 3, 4 }, evs[0].Ids);
        Assert.Equal("length", evs[0].FinishReason);
        Assert.False(w.Inside);

        // A second End() call is a no-op: nothing left in flight.
        Assert.Empty(w.End("length"));
    }

    [Fact]
    public void EndReportsEmptyBodyWhenStreamEndsRightAfterStart()
    {
        var w = new ToolWatcher(SynthMap(), "<tool_call>", "</tool_call>");
        w.Feed(new uint[] { START });
        Assert.True(w.Inside);

        var evs = w.End(); // no finish reason known
        Assert.Single(evs);
        Assert.Equal(WatcherEventKind.Truncated, evs[0].Kind);
        Assert.Empty(evs[0].Ids);
        Assert.Null(evs[0].FinishReason);
    }

    [Fact]
    public void EndOutsideRegionEmitsNothing()
    {
        var w = new ToolWatcher(SynthMap(), "<tool_call>", "</tool_call>");
        w.Feed(new uint[] { START, 3, END, 4 });
        Assert.False(w.Inside);
        Assert.Empty(w.End("stop"));
    }

    // ── Overflow: region buffer cap (defect 2) ─────────────────────────────────
    //
    // The region buffer used to grow without bound: a client that can make
    // the model emit a start marker without a matching end marker could
    // grow it to the entire remaining generation. The cap must be enforced
    // and the overflow must be a defined, observable event, not a silent
    // truncation.

    [Fact]
    public void RegionCapDefaultsAndIsSettable()
    {
        var w = new ToolWatcher(SynthMap(), "<tool_call>", "</tool_call>");
        Assert.Equal(ToolWatcher.DefaultRegionCap, w.RegionCap);

        w.SetRegionCap(3);
        Assert.Equal(3, w.RegionCap);

        // 0 resets to the default rather than becoming an unusable cap.
        w.SetRegionCap(0);
        Assert.Equal(ToolWatcher.DefaultRegionCap, w.RegionCap);

        var w2 = new ToolWatcher(SynthMap(), "<tool_call>", "</tool_call>", regionCap: 3);
        Assert.Equal(3, w2.RegionCap);
    }

    [Fact]
    public void OverflowFiresOnceAtCapThenResyncsOnEndMarker()
    {
        var w = new ToolWatcher(SynthMap(), "<tool_call>", "</tool_call>", regionCap: 3);
        // Region body is 5 tokens long against a cap of 3: must overflow
        // once, with exactly the first 3 tokens, and must NOT also emit a
        // region event for the same region when the end marker eventually
        // arrives.
        var evs = w.Feed(new uint[] { START, 1, 2, 3, 4, 5, END, 9 });
        Assert.Equal(2, evs.Count);
        Assert.Equal(WatcherEventKind.Overflow, evs[0].Kind);
        Assert.Equal(new uint[] { 1, 2, 3 }, evs[0].Ids);
        Assert.Equal(WatcherEventKind.Passthrough, evs[1].Kind);
        Assert.Equal(new uint[] { 9 }, evs[1].Ids);
        Assert.False(w.Inside);
    }

    [Fact]
    public void OverflowThenTruncatedReportsBoth()
    {
        // A region that overflows and then never sees an end marker must
        // report BOTH: the overflow (memory bound hit) and the
        // truncation (stream ended without a close). They are orthogonal
        // signals.
        var w = new ToolWatcher(SynthMap(), "<tool_call>", "</tool_call>", regionCap: 2);
        var evs = w.Feed(new uint[] { START, 1, 2, 3, 4 });
        Assert.Single(evs);
        Assert.Equal(WatcherEventKind.Overflow, evs[0].Kind);
        Assert.Equal(new uint[] { 1, 2 }, evs[0].Ids);

        evs = w.End("length");
        Assert.Single(evs);
        Assert.Equal(WatcherEventKind.Truncated, evs[0].Kind);
        Assert.Equal(new uint[] { 1, 2 }, evs[0].Ids);
        Assert.Equal("length", evs[0].FinishReason);
    }

    [Fact]
    public void ExactCapDoesNotOverflow()
    {
        // Off-by-one check: a region whose body is exactly `cap` tokens
        // must close cleanly as Region, not as Overflow.
        var w = new ToolWatcher(SynthMap(), "<tool_call>", "</tool_call>", regionCap: 3);
        var evs = w.Feed(new uint[] { START, 1, 2, 3, END });
        Assert.Single(evs);
        Assert.Equal(WatcherEventKind.Region, evs[0].Kind);
        Assert.Equal(new uint[] { 1, 2, 3 }, evs[0].Ids);
    }
}

File: packages/java/src/test/java/ai/codec/ToolWatcherTests.java
// SPDX-License-Identifier: MIT
package ai.codec;

import org.junit.jupiter.api.Test;

import java.util.HashMap;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class ToolWatcherTests {
    private static final long START = 90L;
    private static final long END = 91L;

    private static TokenizerMap synthMap() {
        TokenizerMap m = new TokenizerMap();
        m.id = "test/synth";
        m.version = "2";
        m.vocabSize = 100;
        m.encoder = "byte_level";
        m.vocab = new HashMap<>();
        m.vocab.put("hello", 0);
        m.vocab.put("world", 1);
        m.vocab.put("!", 2);
        m.vocab.put("foo", 3);
        m.vocab.put("bar", 4);
        m.specialTokens = new HashMap<>();
        m.specialTokens.put("<tool_call>", 90);
        m.specialTokens.put("</tool_call>", 91);
        return m;
    }

    @Test
    void passthroughThenRegionThenPassthrough() {
        ToolWatcher w = new ToolWatcher(synthMap(), "<tool_call>", "</tool_call>");
        // "hello world <tool_call> foo bar </tool_call> hello !"
        List<WatcherEvent> evs = w.feed(new long[] { 0, 1, START, 3, 4, END, 0, 2 });
        assertEquals(3, evs.size());

        assertEquals(WatcherEventKind.PASSTHROUGH, evs.get(0).getKind());
        assertArrayEquals(new long[] { 0, 1 }, evs.get(0).getIds());

        assertEquals(WatcherEventKind.REGION, evs.get(1).getKind());
        assertArrayEquals(new long[] { 3, 4 }, evs.get(1).getIds());

        assertEquals(WatcherEventKind.PASSTHROUGH, evs.get(2).getKind());
        assertArrayEquals(new long[] { 0, 2 }, evs.get(2).getIds());

        assertFalse(w.isInside());
    }

    @Test
    void regionSplitAcrossFeeds() {
        ToolWatcher w = new ToolWatcher(synthMap(), "<tool_call>", "</tool_call>");

        // Feed 1: region opens but doesn't close.
        List<WatcherEvent> evs = w.feed(new long[] { 0, START, 3 });
        assertEquals(1, evs.size());
        assertEquals(WatcherEventKind.PASSTHROUGH, evs.get(0).getKind());
        assertArrayEquals(new long[] { 0 }, evs.get(0).getIds());
        assertTrue(w.isInside());

        // Feed 2: closes the region with body accumulated across both feeds.
        evs = w.feed(new long[] { 4, END, 1 });
        assertEquals(2, evs.size());
        assertEquals(WatcherEventKind.REGION, evs.get(0).getKind());
        assertArrayEquals(new long[] { 3, 4 }, evs.get(0).getIds());
        assertEquals(WatcherEventKind.PASSTHROUGH, evs.get(1).getKind());
        assertArrayEquals(new long[] { 1 }, evs.get(1).getIds());
        assertFalse(w.isInside());
    }

    @Test
    void multipleRegionsInOneFeed() {
        ToolWatcher w = new ToolWatcher(synthMap(), "<tool_call>", "</tool_call>");
        List<WatcherEvent> evs = w.feed(new long[] { 0, START, 3, END, 1, START, 4, END, 2 });
        assertEquals(5, evs.size());
        assertEquals(WatcherEventKind.PASSTHROUGH, evs.get(0).getKind());
        assertEquals(WatcherEventKind.REGION, evs.get(1).getKind());
        assertArrayEquals(new long[] { 3 }, evs.get(1).getIds());
        assertEquals(WatcherEventKind.PASSTHROUGH, evs.get(2).getKind());
        assertEquals(WatcherEventKind.REGION, evs.get(3).getKind());
        assertArrayEquals(new long[] { 4 }, evs.get(3).getIds());
        assertEquals(WatcherEventKind.PASSTHROUGH, evs.get(4).getKind());
    }

    @Test
    void strayEndPassesThrough() {
        ToolWatcher w = new ToolWatcher(synthMap(), "<tool_call>", "</tool_call>");
        List<WatcherEvent> evs = w.feed(new long[] { 0, END, 1 });
        // End with no preceding start: ordinary token.
        assertEquals(1, evs.size());
        assertEquals(WatcherEventKind.PASSTHROUGH, evs.get(0).getKind());
        assertArrayEquals(new long[] { 0, END, 1 }, evs.get(0).getIds());
    }

    @Test
    void missingSpecialNameThrows() {
        assertThrows(ToolWatcherException.class,
                () -> new ToolWatcher(synthMap(), "<not_real>", "</tool_call>"));
    }

    @Test
    void resetDropsInFlightRegion() {
        ToolWatcher w = new ToolWatcher(synthMap(), "<tool_call>", "</tool_call>");
        w.feed(new long[] { START, 3, 4 });
        assertTrue(w.isInside());
        w.reset();
        assertFalse(w.isInside());
        // End marker now becomes a stray (no buffered body).
        List<WatcherEvent> evs = w.feed(new long[] { END, 1 });
        assertEquals(1, evs.size());
        assertEquals(WatcherEventKind.PASSTHROUGH, evs.get(0).getKind());
        assertArrayEquals(new long[] { END, 1 }, evs.get(0).getIds());
    }

    /**
     * No-decode contract: must operate on raw IDs only, never reading map.vocab.
     * Use a map with empty vocab and feed IDs outside any plausible vocab range.
     * The watcher must emit them verbatim.
     */
    @Test
    void neverDecodesOperatesOnRawIds() {
        TokenizerMap noVocab = new TokenizerMap();
        noVocab.id = "test/no-vocab";
        noVocab.version = "2";
        noVocab.vocabSize = 4;
        noVocab.encoder = "byte_level";
        noVocab.vocab = new HashMap<>();
        noVocab.specialTokens = new HashMap<>();
        noVocab.specialTokens.put("<tool_call>", 90);
        noVocab.specialTokens.put("</tool_call>", 91);

        ToolWatcher w = new ToolWatcher(noVocab, "<tool_call>", "</tool_call>");
        long bigA = 0xFFFFFF00L;
        long bigB = 0xDEADBEEFL;
        long bigC = 0xCAFEBABEL;
        List<WatcherEvent> evs = w.feed(new long[] { 12345L, bigA, START, bigB, bigC, END, 99999L });
        assertEquals(3, evs.size());
        assertArrayEquals(new long[] { 12345L, bigA }, evs.get(0).getIds());
        assertArrayEquals(new long[] { bigB, bigC }, evs.get(1).getIds()); // body verbatim
        assertArrayEquals(new long[] { 99999L }, evs.get(2).getIds());
    }

    @Test
    void intOverloadAcceptsWireFrameType() {
        // CodecFrame.ids is int[]; verify the int overload works.
        ToolWatcher w = new ToolWatcher(synthMap(), "<tool_call>", "</tool_call>");
        List<WatcherEvent> evs = w.feed(new int[] { 0, 1, (int) START, 3, (int) END, 2 });
        assertEquals(3, evs.size());
        assertEquals(WatcherEventKind.REGION, evs.get(1).getKind());
        assertArrayEquals(new long[] { 3 }, evs.get(1).getIds());
    }

    // ── Ordering: interleaved events in stream order (defect 3) ──────────────
    //
    // [a, S, X, E, b, S, Y, E, c] must produce five ORDERED events:
    // passthrough(a) / region(X) / passthrough(b) / region(Y) / passthrough(c).
    // This is the exact shape every language's watcher must agree on.

    @Test
    void orderingMatchesDefect3Example() {
        ToolWatcher w = new ToolWatcher(synthMap(), "<tool_call>", "</tool_call>");
        long a = 0, b = 1, c = 2, x = 3, y = 4; // hello, world, !, foo, bar
        List<WatcherEvent> evs = w.feed(new long[] { a, START, x, END, b, START, y, END, c });
        assertEquals(5, evs.size());
        assertEquals(WatcherEventKind.PASSTHROUGH, evs.get(0).getKind());
        assertArrayEquals(new long[] { a }, evs.get(0).getIds());
        assertEquals(WatcherEventKind.REGION, evs.get(1).getKind());
        assertArrayEquals(new long[] { x }, evs.get(1).getIds());
        assertEquals(WatcherEventKind.PASSTHROUGH, evs.get(2).getKind());
        assertArrayEquals(new long[] { b }, evs.get(2).getIds());
        assertEquals(WatcherEventKind.REGION, evs.get(3).getKind());
        assertArrayEquals(new long[] { y }, evs.get(3).getIds());
        assertEquals(WatcherEventKind.PASSTHROUGH, evs.get(4).getKind());
        assertArrayEquals(new long[] { c }, evs.get(4).getIds());
    }

    // ── Nested start markers (defect 5) ───────────────────────────────────────

    @Test
    void nestedStartIsDroppedFromBodyButObservable() {
        ToolWatcher w = new ToolWatcher(synthMap(), "<tool_call>", "</tool_call>");
        // S 0 S 1 E 2 -> NESTED_START / REGION([0,1]) / PASSTHROUGH([2])
        List<WatcherEvent> evs = w.feed(new long[] { START, 0, START, 1, END, 2 });
        assertEquals(3, evs.size());
        assertEquals(WatcherEventKind.NESTED_START, evs.get(0).getKind());
        assertArrayEquals(new long[] { START }, evs.get(0).getIds());
        assertEquals(WatcherEventKind.REGION, evs.get(1).getKind());
        assertArrayEquals(new long[] { 0, 1 }, evs.get(1).getIds());
        assertEquals(WatcherEventKind.PASSTHROUGH, evs.get(2).getKind());
        assertArrayEquals(new long[] { 2 }, evs.get(2).getIds());
    }

    // ── Truncation: end() while inside a region (defect 1) ────────────────────
    //
    // An unterminated region (stream ends mid tool-call, e.g. the model hit
    // its length limit) used to be silently dropped: no event, no signal,
    // indistinguishable from a model that never called a tool. end() must
    // report it, carrying the finish reason so a length stop is
    // distinguishable from a malformed emission.

    @Test
    void endEmitsTruncatedWithFinishReason() {
        ToolWatcher w = new ToolWatcher(synthMap(), "<tool_call>", "</tool_call>");
        List<WatcherEvent> evs = w.feed(new long[] { 0, START, 3, 4 });
        assertEquals(1, evs.size());
        assertEquals(WatcherEventKind.PASSTHROUGH, evs.get(0).getKind());
        assertTrue(w.isInside());

        evs = w.end("length");
        assertEquals(1, evs.size());
        assertEquals(WatcherEventKind.TRUNCATED, evs.get(0).getKind());
        assertArrayEquals(new long[] { 3, 4 }, evs.get(0).getIds());
        assertEquals("length", evs.get(0).getFinishReason());
        assertFalse(w.isInside());

        // A second end() call is a no-op: nothing left in flight.
        assertTrue(w.end("length").isEmpty());
    }

    @Test
    void endReportsEmptyBodyWhenStreamEndsRightAfterStart() {
        ToolWatcher w = new ToolWatcher(synthMap(), "<tool_call>", "</tool_call>");
        w.feed(new long[] { START });
        assertTrue(w.isInside());

        List<WatcherEvent> evs = w.end(); // no finish reason known
        assertEquals(1, evs.size());
        assertEquals(WatcherEventKind.TRUNCATED, evs.get(0).getKind());
        assertArrayEquals(new long[0], evs.get(0).getIds());
        assertNull(evs.get(0).getFinishReason());
    }

    @Test
    void endOutsideRegionEmitsNothing() {
        ToolWatcher w = new ToolWatcher(synthMap(), "<tool_call>", "</tool_call>");
        w.feed(new long[] { START, 3, END, 4 });
        assertFalse(w.isInside());
        assertTrue(w.end("stop").isEmpty());
    }

    // ── Overflow: region buffer cap (defect 2) ─────────────────────────────────
    //
    // The region buffer used to grow without bound: a client that can make
    // the model emit a start marker without a matching end marker could
    // grow it to the entire remaining generation. The cap must be enforced
    // and the overflow must be a defined, observable event, not a silent
    // truncation.

    @Test
    void regionCapDefaultsAndIsSettable() {
        ToolWatcher w = new ToolWatcher(synthMap(), "<tool_call>", "</tool_call>");
        assertEquals(ToolWatcher.DEFAULT_REGION_CAP, w.getRegionCap());

        w.setRegionCap(3);
        assertEquals(3, w.getRegionCap());

        // 0 resets to the default rather than becoming an unusable cap.
        w.setRegionCap(0);
        assertEquals(ToolWatcher.DEFAULT_REGION_CAP, w.getRegionCap());

        ToolWatcher w2 = new ToolWatcher(synthMap(), "<tool_call>", "</tool_call>", 3);
        assertEquals(3, w2.getRegionCap());
    }

    @Test
    void overflowFiresOnceAtCapThenResyncsOnEndMarker() {
        ToolWatcher w = new ToolWatcher(synthMap(), "<tool_call>", "</tool_call>", 3);
        // Region body is 5 tokens long against a cap of 3: must overflow
        // once, with exactly the first 3 tokens, and must NOT also emit
        // a region event for the same region when the end marker
        // eventually arrives.
        List<WatcherEvent> evs = w.feed(new long[] { START, 1, 2, 3, 4, 5, END, 9 });
        assertEquals(2, evs.size());
        assertEquals(WatcherEventKind.OVERFLOW, evs.get(0).getKind());
        assertArrayEquals(new long[] { 1, 2, 3 }, evs.get(0).getIds());
        assertEquals(WatcherEventKind.PASSTHROUGH, evs.get(1).getKind());
        assertArrayEquals(new long[] { 9 }, evs.get(1).getIds());
        assertFalse(w.isInside());
    }

    @Test
    void overflowThenTruncatedReportsBoth() {
        // A region that overflows and then never sees an end marker must
        // report BOTH: the overflow (memory bound hit) and the
        // truncation (stream ended without a close). They are orthogonal
        // signals.
        ToolWatcher w = new ToolWatcher(synthMap(), "<tool_call>", "</tool_call>", 2);
        List<WatcherEvent> evs = w.feed(new long[] { START, 1, 2, 3, 4 });
        assertEquals(1, evs.size());
        assertEquals(WatcherEventKind.OVERFLOW, evs.get(0).getKind());
        assertArrayEquals(new long[] { 1, 2 }, evs.get(0).getIds());

        evs = w.end("length");
        assertEquals(1, evs.size());
        assertEquals(WatcherEventKind.TRUNCATED, evs.get(0).getKind());
        assertArrayEquals(new long[] { 1, 2 }, evs.get(0).getIds());
        assertEquals("length", evs.get(0).getFinishReason());
    }

    @Test
    void exactCapDoesNotOverflow() {
        // Off-by-one check: a region whose body is exactly `cap` tokens
        // must close cleanly as REGION, not as OVERFLOW.
        ToolWatcher w = new ToolWatcher(synthMap(), "<tool_call>", "</tool_call>", 3);
        List<WatcherEvent> evs = w.feed(new long[] { START, 1, 2, 3, END });
        assertEquals(1, evs.size());
        assertEquals(WatcherEventKind.REGION, evs.get(0).getKind());
        assertArrayEquals(new long[] { 1, 2, 3 }, evs.get(0).getIds());
    }
}
